From 4eca366b66953e15b1244351d1dc1054e6bc586e Mon Sep 17 00:00:00 2001
From: Xinhui Yang <cyan@cyano.uk>
Date: Mon, 11 Aug 2025 10:42:41 +0800
Subject: [PATCH 49/57] grub.d/10_linux: Use an in-house script for generating
 linux entries

Now we have to deal with different kernel page sizes, since we want to
show up in the top-level menu, instead of in the "Advanced options for"
submenu.

Also, we want to prefer a certain kernel with specific page size over
the other ones, otherwise the machine might not boot.
---
 util/grub.d/10_linux.in | 613 +++++++++++++++++++++-------------------
 1 file changed, 324 insertions(+), 289 deletions(-)

diff --git a/util/grub.d/10_linux.in b/util/grub.d/10_linux.in
index e896c0c1a..10df977cd 100644
--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -1,315 +1,350 @@
-#! /bin/sh
-set -e
-
-# grub-mkconfig helper script.
-# Copyright (C) 2006,2007,2008,2009,2010  Free Software Foundation, Inc.
-#
-# GRUB is free software: you can redistribute it and/or modify
-# it under the terms of the GNU General Public License as published by
-# the Free Software Foundation, either version 3 of the License, or
-# (at your option) any later version.
-#
-# GRUB is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
-
-prefix="@prefix@"
-exec_prefix="@exec_prefix@"
-datarootdir="@datarootdir@"
-
-. "$pkgdatadir/grub-mkconfig_lib"
-
-export TEXTDOMAIN=@PACKAGE@
-export TEXTDOMAINDIR="@localedir@"
+#!/bin/bash
+
+[ -e /etc/default/grub ] && source /etc/default/grub
+[ -e /etc/default/grub-aosc ] && source /etc/default/grub-aosc
+. /usr/share/grub/grub-mkconfig_lib
+
+indent=""
+tab=$'\t'
+arch=$(uname -m)
+KERNELS=()
+# Using an associative array to keep track of unprocessed kernels.
+# They can be easily modified, unlike ordinary arrays.
+declare -A UNPROCESSED_VERSIONS
+KERNEL_VERSIONS=()
+OS=$(gettext_printf "AOSC OS")
+OS_ID="aosc"
+KERNEL_DIR=${KERNEL_DIR:-/boot}
+
+shopt -s extglob
+
+bool() {
+	[[ "$1" == "true" || "$1" == "1" || "$1" =~ [yY] ]]
+}
 
-CLASS="--class gnu-linux --class gnu --class os"
+info() {
+	echo "-- $@" >&2
+}
 
-if [ "x${GRUB_DISTRIBUTOR}" = "x" ] ; then
-  OS="$(sed 's, release .*$,,g' /etc/system-release)"
+if bool "$AOSC_GRUB_DEBUG" ; then
+	#set -x
+	debug() {
+		echo -e ">> $@" >&2
+	}
 else
-  OS="${GRUB_DISTRIBUTOR}"
-  CLASS="--class $(echo ${GRUB_DISTRIBUTOR} | tr 'A-Z' 'a-z' | cut -d' ' -f1|LC_ALL=C sed 's,[^[:alnum:]_],_,g') ${CLASS}"
+	debug() {
+		: noop
+	}
 fi
 
-# loop-AES arranges things so that /dev/loop/X can be our root device, but
-# the initrds that Linux uses don't like that.
-case ${GRUB_DEVICE} in
-  /dev/loop/*|/dev/loop[0-9])
-    GRUB_DEVICE=`losetup ${GRUB_DEVICE} | sed -e "s/^[^(]*(\([^)]\+\)).*/\1/"`
-  ;;
-esac
-
-: ${GRUB_CMDLINE_LINUX_RECOVERY:=single}
+if [ -z "$GRUB_DEVICE" ] ; then
+	GRUB_DEVICE="`${grub_probe} --target=device /`"
+	GRUB_DEVICE_UUID="`${grub_probe} --device ${GRUB_DEVICE} --target=fs_uuid 2> /dev/null`" || true
+	GRUB_DEVICE_PARTUUID="`${grub_probe} --device ${GRUB_DEVICE} --target=partuuid 2> /dev/null`" || true
+fi
 
-# Default to disabling partition uuid support to maintian compatibility with
-# older kernels.
-: ${GRUB_DISABLE_LINUX_PARTUUID=true}
+if [ -z "$GRUB_DEVICE_BOOT" ] ; then
+	# Device containing our /boot partition.  Usually the same as GRUB_DEVICE.
+	GRUB_DEVICE_BOOT="`${grub_probe} --target=device /boot`"
+	GRUB_DEVICE_BOOT_UUID="`${grub_probe} --device ${GRUB_DEVICE_BOOT} --target=fs_uuid 2> /dev/null`" || true
+fi
 
-# btrfs may reside on multiple devices. We cannot pass them as value of root= parameter
-# and mounting btrfs requires user space scanning, so force UUID in this case.
-if ( [ "x${GRUB_DEVICE_UUID}" = "x" ] && [ "x${GRUB_DEVICE_PARTUUID}" = "x" ] ) \
-    || ( [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ] \
-	&& [ "x${GRUB_DISABLE_LINUX_PARTUUID}" = "xtrue" ] ) \
-    || ( ! test -e "/dev/disk/by-uuid/${GRUB_DEVICE_UUID}" \
-	&& ! test -e "/dev/disk/by-partuuid/${GRUB_DEVICE_PARTUUID}" ) \
-    || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm ); then
-  LINUX_ROOT_DEVICE=${GRUB_DEVICE}
-elif [ "x${GRUB_DEVICE_UUID}" = "x" ] \
-    || [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ]; then
-  LINUX_ROOT_DEVICE=PARTUUID=${GRUB_DEVICE_PARTUUID}
-else
-  LINUX_ROOT_DEVICE=UUID=${GRUB_DEVICE_UUID}
+if [[ -z "$GRUB_DEVICE" || -z "$GRUB_DEVICE_BOOT" ]] ; then
+	gettext_printf "Internal error - Failed to gather information of the root partition\n"
+	exit 1
 fi
 
-case x"$GRUB_FS" in
-    xbtrfs)
+# From GRUB: specify subvolume ID for btrfs, if the filesystem has subvolumes.
+if [ "$GRUB_FS" = "btrfs" ] ; then
 	rootsubvol="`make_system_path_relative_to_its_root /`"
 	rootsubvol="${rootsubvol#/}"
 	if [ "x${rootsubvol}" != x ]; then
-	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
-	fi;;
-    xzfs)
-	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || true`
-	bootfs="`make_system_path_relative_to_its_root / | sed -e "s,@$,,"`"
-	LINUX_ROOT_DEVICE="ZFS=${rpool}${bootfs%/}"
-	;;
-esac
-
-title_correction_code=
-
-linux_entry ()
-{
-  os="$1"
-  version="$2"
-  type="$3"
-  args="$4"
-
-  if [ -z "$boot_device_id" ]; then
-      boot_device_id="$(grub_get_device_id "${GRUB_DEVICE}")"
-  fi
-  if [ x$type != xsimple ] ; then
-      case $type in
-	  recovery)
-	      title="$(gettext_printf "%s, with Linux %s (recovery mode)" "${os}" "${version}")" ;;
-	  *)
-	      title="$(gettext_printf "%s, with Linux %s" "${os}" "${version}")" ;;
-      esac
-      if [ x"$title" = x"$GRUB_ACTUAL_DEFAULT" ] || [ x"Previous Linux versions>$title" = x"$GRUB_ACTUAL_DEFAULT" ]; then
-	  replacement_title="$(echo "Advanced options for ${OS}" | sed 's,>,>>,g')>$(echo "$title" | sed 's,>,>>,g')"
-	  quoted="$(echo "$GRUB_ACTUAL_DEFAULT" | grub_quote)"
-	  title_correction_code="${title_correction_code}if [ \"x\$default\" = '$quoted' ]; then default='$(echo "$replacement_title" | grub_quote)'; fi;"
-	  grub_warn "$(gettext_printf "Please don't use old title \`%s' for GRUB_DEFAULT, use \`%s' (for versions before 2.00) or \`%s' (for 2.00 or later)" "$GRUB_ACTUAL_DEFAULT" "$replacement_title" "gnulinux-advanced-$boot_device_id>gnulinux-$version-$type-$boot_device_id")"
-      fi
-      echo "menuentry '$(echo "$title" | grub_quote)' ${CLASS} \$menuentry_id_option 'gnulinux-$version-$type-$boot_device_id' {" | sed "s/^/$submenu_indentation/"
-  else
-      echo "menuentry '$(echo "$os" | grub_quote)' ${CLASS} \$menuentry_id_option 'gnulinux-simple-$boot_device_id' {" | sed "s/^/$submenu_indentation/"
-  fi      
-  if [ x$type != xrecovery ] ; then
-      save_default_entry | grub_add_tab
-  fi
-
-  # Use ELILO's generic "efifb" when it's known to be available.
-  # FIXME: We need an interface to select vesafb in case efifb can't be used.
-  if [ "x$GRUB_GFXPAYLOAD_LINUX" = x ]; then
-      echo "	load_video" | sed "s/^/$submenu_indentation/"
-      if grep -qx "CONFIG_FB_EFI=y" "${config}" 2> /dev/null \
-	  && grep -qx "CONFIG_VT_HW_CONSOLE_BINDING=y" "${config}" 2> /dev/null; then
-	  echo "	set gfxpayload=keep" | sed "s/^/$submenu_indentation/"
-      fi
-  else
-      if [ "x$GRUB_GFXPAYLOAD_LINUX" != xtext ]; then
-	  echo "	load_video" | sed "s/^/$submenu_indentation/"
-      fi
-      echo "	set gfxpayload=$GRUB_GFXPAYLOAD_LINUX" | sed "s/^/$submenu_indentation/"
-  fi
-
-  echo "	insmod gzio" | sed "s/^/$submenu_indentation/"
-
-  if [ x$dirname = x/ ]; then
-    if [ -z "${prepare_root_cache}" ]; then
-      prepare_root_cache="$(prepare_grub_to_access_device ${GRUB_DEVICE} | grub_add_tab)"
-    fi
-    printf '%s\n' "${prepare_root_cache}" | sed "s/^/$submenu_indentation/"
-  else
-    if [ -z "${prepare_boot_cache}" ]; then
-      prepare_boot_cache="$(prepare_grub_to_access_device ${GRUB_DEVICE_BOOT} | grub_add_tab)"
-    fi
-    printf '%s\n' "${prepare_boot_cache}" | sed "s/^/$submenu_indentation/"
-  fi
-  message="$(gettext_printf "Loading Linux %s ..." ${version})"
-  sed "s/^/$submenu_indentation/" << EOF
-	echo	'$(echo "$message" | grub_quote)'
-	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args}
-EOF
-  if test -n "${initrd}" ; then
-    # TRANSLATORS: ramdisk isn't identifier. Should be translated.
-    message="$(gettext_printf "Loading initial ramdisk ...")"
-    initrd_path=
-    for i in ${initrd}; do
-      initrd_path="${initrd_path} ${rel_dirname}/${i}"
-    done
-    sed "s/^/$submenu_indentation/" << EOF
-	echo	'$(echo "$message" | grub_quote)'
-	initrd	$(echo $initrd_path)
-EOF
-  fi
-  sed "s/^/$submenu_indentation/" << EOF
-}
-EOF
-}
-
-machine=`uname -m`
-case "x$machine" in
-    xi?86 | xx86_64)
-	list=
-	for i in /boot/vmlinuz-* /vmlinuz-* /boot/kernel-* ; do
-	    if grub_file_is_not_garbage "$i" ; then list="$list $i" ; fi
-	done ;;
-    *) 
-	list=
-	for i in /boot/vmlinuz-* /boot/vmlinux-* /vmlinuz-* /vmlinux-* /boot/kernel-* ; do
-                  if grub_file_is_not_garbage "$i" ; then list="$list $i" ; fi
-	done ;;
-esac
-
-case "$machine" in
-    i?86) GENKERNEL_ARCH="x86" ;;
-    mips|mips64) GENKERNEL_ARCH="mips" ;;
-    mipsel|mips64el) GENKERNEL_ARCH="mipsel" ;;
-    arm*) GENKERNEL_ARCH="arm" ;;
-    *) GENKERNEL_ARCH="$machine" ;;
-esac
-
-prepare_boot_cache=
-prepare_root_cache=
-boot_device_id=
-title_correction_code=
-
-# Extra indentation to add to menu entries in a submenu. We're not in a submenu
-# yet, so it's empty. In a submenu it will be equal to '\t' (one tab).
-submenu_indentation=""
-
-# Perform a reverse version sort on the entire list.
-# Temporarily replace the '.old' suffix by ' 1' and append ' 2' for all
-# other files to order the '.old' files after their non-old counterpart
-# in reverse-sorted order.
-
-reverse_sorted_list=$(echo $list | tr ' ' '\n' | sed -e 's/\.old$/ 1/; / 1$/! s/$/ 2/' | version_sort -r | sed -e 's/ 1$/.old/; s/ 2$//')
-
-if [ "x$GRUB_TOP_LEVEL" != x ]; then
-  reverse_sorted_list=$(grub_move_to_front "$GRUB_TOP_LEVEL" ${reverse_sorted_list})
+		GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
+	fi
+fi
+# Depends on whether /boot is a mounted filesystem, the path GRUB loads the
+# kernel changes.
+if [ "$GRUB_DEVICE" == "$GRUB_DEVICE_BOOT" ] ; then
+	GRUB_KERNEL_DIR="${KERNEL_DIR%%/}/"
+else
+	GRUB_KERNEL_DIR="/"
 fi
 
-is_top_level=true
-for linux in ${reverse_sorted_list}; do
-  gettext_printf "Found linux image: %s\n" "$linux" >&2
-  basename=`basename $linux`
-  dirname=`dirname $linux`
-  rel_dirname=`make_system_path_relative_to_its_root $dirname`
-  version=`echo $basename | sed -e "s,^[^0-9]*-,,g"`
-  alt_version=`echo $version | sed -e "s,\.old$,,g"`
-  linux_root_device_thisversion="${LINUX_ROOT_DEVICE}"
-
-  initrd_early=
-  for i in ${GRUB_EARLY_INITRD_LINUX_STOCK} \
-	   ${GRUB_EARLY_INITRD_LINUX_CUSTOM}; do
-    if test -e "${dirname}/${i}" ; then
-      initrd_early="${initrd_early} ${i}"
-    fi
-  done
-
-  initrd_real=
-  for i in "initrd.img-${version}" "initrd-${version}.img" \
-	   "initrd-${alt_version}.img.old" "initrd-${version}.gz" \
-	   "initrd-${alt_version}.gz.old" "initrd-${version}" \
-	   "initramfs-${version}.img" "initramfs-${alt_version}.img.old" \
-	   "initrd.img-${alt_version}" "initrd-${alt_version}.img" \
-	   "initrd-${alt_version}" "initramfs-${alt_version}.img" \
-	   "initramfs-genkernel-${version}" \
-	   "initramfs-genkernel-${alt_version}" \
-	   "initramfs-genkernel-${GENKERNEL_ARCH}-${version}" \
-	   "initramfs-genkernel-${GENKERNEL_ARCH}-${alt_version}"; do
-    if test -e "${dirname}/${i}" ; then
-      initrd_real="${i}"
-      break
-    fi
-  done
+MENUENTRY_PREP="$(prepare_grub_to_access_device "$GRUB_DEVICE_BOOT")"
 
-  initrd=
-  if test -n "${initrd_early}" || test -n "${initrd_real}"; then
-    initrd="${initrd_early} ${initrd_real}"
+if [ -z "$GRUB_GFXPAYLOAD_LINUX" ] ; then
+	_GFXPAYLOAD_LINUX="set gfxpayload=keep"
+else
+	_GFXPAYLOAD_LINUX="set gfxpayload=$GRUB_GFXPAYLOAD_LINUX"
+fi
 
-    initrd_display=
-    for i in ${initrd}; do
-      initrd_display="${initrd_display} ${dirname}/${i}"
-    done
-    gettext_printf "Found initrd image: %s\n" "$(echo $initrd_display)" >&2
-  fi
+# Handle microcodes.
+for rd in ${GRUB_EARLY_INITRD_LINUX_STOCK[@]} \
+	${GRUB_EARLY_INITRD_LINUX_CUSTOM[@]} ; do
+	if [ -e "$KERNEL_DIR"/"$rd" ] ; then
+		INITRD_EARLY="${INITRD_EARLY} ${rd}"
+	fi
+done
+# Returns the default page size suffix.
+# WARN $arch is the output of `uname -m`, which is not exactly the same as
+# `dpkg --print-architecture`.
+default_pagesize() {
+	case "$arch" in
+		loongarch64)
+			printf "16k"
+			;;
+		*)
+			printf "4k"
+			;;
+	esac
+}
 
-  config=
-  for i in "${dirname}/config-${version}" "${dirname}/config-${alt_version}" "/etc/kernels/kernel-config-${version}" ; do
-    if test -e "${i}" ; then
-      config="${i}"
-      break
-    fi
-  done
+pagesize_str() {
+	case "$1" in
+		4k)
+			if [ "$(default_pagesize)" = "$1" ] ; then
+				printf ""
+			else
+				gettext_printf "4K Page Size"
+			fi
+			;;
+		16k)
+			if [ "$(default_pagesize)" = "$1" ] ; then
+				printf ""
+			else
+				gettext_printf "16K Page Size"
+			fi
+			;;
+		64k)
+			if [ "$(default_pagesize)" = "$1" ] ; then
+				printf ""
+			else
+				gettext_printf "64K Page Size"
+			fi
+			;;
+		*)
+			# Unknown page size.
+			echo ""
+			;;
+	esac
+}
 
-  initramfs=
-  if test -n "${config}" ; then
-      initramfs=`grep CONFIG_INITRAMFS_SOURCE= "${config}" | cut -f2 -d= | tr -d \"`
-  fi
+# Localize the given variant.
+variant_str() {
+	case "$1" in
+		rc)
+			gettext_printf "RC Kernel"
+			;;
+		lts)
+			gettext_printf "LTS Kernel"
+			;;
+		main)
+			: noop
+			;;
+		*)
+			gettext_printf "Kernel variant '%s'" "$1"
+			;;
+	esac
+}
 
-  if ( test -z "${initramfs}" && test -z "${initrd_real}" ) || \
-    [[ "${initrd_real}" = "intel-ucode.img" ]]; then
-    # "UUID=" and "ZFS=" magic is parsed by initrd or initramfs.  Since there's
-    # no initrd or builtin initramfs, it can't work here.
-    if [ "x${GRUB_DEVICE_PARTUUID}" = "x" ] \
-	|| [ "x${GRUB_DISABLE_LINUX_PARTUUID}" = "xtrue" ]; then
+# Break localversion apart, return the processed comments.
+# E.g.
+# - "-aosc-main" -> ""
+# - "-aosc-lts" -> " (LTS Kernel)"
+# - "-aosc-main-4k" -> " (4K Page Size)"
+# - "-aosc-rc-4k" -> " (RC Kernel, 4K Page Size)"
+get_comments() {
+	local _str="$1" _fullver="$1" segs=() comments=() sep=", " variant pagesize
+	# Not an AOSC OS kernel
+	if ! [[ "$_str" =~ \.[0-9]+-aosc- ]] ; then
+		return
+	fi
+
+	_str="${_str##*-aosc-}"
+	IFS=-
+	segs=($_str)
+	unset IFS
+	variant="$(variant_str "${segs[0]}")"
+	pagesize="$(pagesize_str "${segs[1]}")"
+	comment="${variant}${sep}${pagesize}"
+	if [ "x$comment" = "x$sep" ] ; then
+		return
+	fi
+	comment="${comment##$sep}"
+	comment="${comment%%$sep}"
+	printf "%s" " ($comment)"
+}
 
-	linux_root_device_thisversion=${GRUB_DEVICE}
-    else
-	linux_root_device_thisversion=PARTUUID=${GRUB_DEVICE_PARTUUID}
-    fi
-  fi
+linux_entry() {
+	title="$1"
+	kernel="$2"
+	initrd="$3"
+	args="$4"
+	_id="$5"
+	
+	printf "${indent}menuentry '%s' --class linux \$menuentry_id_option '%s' {\n" "$title" "$_id"
+	printf "$(echo "$MENUENTRY_PREP" | sed "s|^|\t${indent}|g")\n"
+	printf "${indent}\t$_GFXPAYLOAD_LINUX\n"
+	printf "${indent}\tinsmod	gzio\n"
+	printf "${indent}\techo	'%s'\n" "$(gettext_printf "Loading kernel $kernel ...")"
+	printf "${indent}\tlinux	%s %s\n" "$kernel" "$args"
+	if [ -n "$initrd" ] ; then
+		printf "${indent}\techo	'%s'\n" "$(gettext_printf "Loading initial ramdisk ...")"
+		printf "${indent}\tinitrd	%s\n" "$initrd"
+	fi
+	printf "${indent}\techo	'%s'\n" "$(gettext_printf "Booting up $OS ...")"
+	printf "${indent}}\n\n"
+}
 
-  # The GRUB_DISABLE_SUBMENU option used to be different than others since it was
-  # mentioned in the documentation that has to be set to 'y' instead of 'true' to
-  # enable it. This caused a lot of confusion to users that set the option to 'y',
-  # 'yes' or 'true'. This was fixed but all of these values must be supported now.
-  if [ "x${GRUB_DISABLE_SUBMENU}" = xyes ] || [ "x${GRUB_DISABLE_SUBMENU}" = xy ]; then
-    GRUB_DISABLE_SUBMENU="true"
-  fi
+find_rootdev() {
+	local ident_type=uuid
+	if ! [ -e "$GRUB_DEVICE" ] ; then
+		ident_type=error
+	elif bool "$GRUB_DISABLE_LINUX_UUID" ; then
+		if bool "$GRUB_DISABLE_LINUX_PARTUUID" ; then
+			ident_type=path
+		else
+			ident_type=partuuid
+		fi
+	else
+		if [ ! -e "/dev/disk/by-uuid/${GRUB_DEVICE_UUID,,}" ] ; then
+			ident_type=partuuid
+		fi
+		if [ ! -e "/dev/disk/by-partuuid/${GRUB_DEVICE_PARTUUID,,}" ] || \
+			bool "${GRUB_DISABLE_LINUX_PARTUUID}" || \
+			uses_abstraction "${GRUB_DEVICE}" lvm ; then
+			ident_type=path
+		fi
+
+	fi
+	case "$ident_type" in
+		path)
+			printf "%s" "$GRUB_DEVICE"
+			;;
+		uuid)
+			printf "UUID=%s" "$GRUB_DEVICE_UUID"
+			return
+			;;
+		partuuid)
+			printf "PARTUUID=%s" "$GRUB_DEVICE_PARTUUID"
+			;;
+		*)
+			# Propogate the error
+			;;
+	esac
+}
 
-  if [ "x$is_top_level" = xtrue ] && [ "x${GRUB_DISABLE_SUBMENU}" != xtrue ]; then
-    linux_entry "${OS}" "${version}" simple \
-    "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
+# Prioritize kernel names in "$@", based on the default size.
+# The kernel with the default size will be always on top. If it does
+# not exist, then the kernel without page size suffix will be on top.
+prioritize() {
+	local kernels newkernels default_pgsize flag
+	kernels=("$@")
+	newkernels=()
+	default_pgsize="$(default_pagesize)"
+	for k in "${kernels[@]}" ; do
+		# This system probably don't have other page sizes.
+		if [ "${k%%-[0-9]*k}" = "$k" ] ; then
+			newkernels=("$k" "${newkernels[@]}")
+			continue
+		fi
+		if [ "${k%%-$default_pgsize}" != "$k" ] ; then
+			newkernels=("$k" "${newkernels[@]}")
+			continue
+		fi
+		newkernels+=("$k")
+	done
+	OLDIFS="$IFS"
+	IFS=$'\n'
+	echo -n "${newkernels[*]}"
+	IFS="$OLDIFS"
+}
 
-    submenu_indentation="$grub_tab"
-    
-    if [ -z "$boot_device_id" ]; then
-	boot_device_id="$(grub_get_device_id "${GRUB_DEVICE}")"
-    fi
-    # TRANSLATORS: %s is replaced with an OS name
-    echo "submenu '$(gettext_printf "Advanced options for %s" "${OS}" | grub_quote)' \$menuentry_id_option 'gnulinux-advanced-$boot_device_id' {"
-    is_top_level=false
-  fi
+scan_for_kernels() {
+	IFS=$'\n'
+	KERNELS=($(find $KERNEL_DIR -maxdepth 1 -type f -a \( -name 'vmlinux-*' -o -name 'vmlinuz-*' -o -name 'kernel-*' \) -printf '%P\n'))
+	if [ "${#KERNELS[@]}" -lt 1 ] ; then
+		unset IFS
+		return
+	fi
+	KERNEL_VERSIONS+=($(echo -e "${KERNELS[*]}" | grep -o '[0-9]*\.[0-9]*\.[0-9]*' | sort -rVu))
+	for ver in "${KERNEL_VERSIONS[@]}" ; do
+		UNPROCESSED_VERSIONS["$ver"]="$ver"
+	done
+	debug "Kernel versions:\n${KERNEL_VERSIONS[*]}"
+	unset IFS k
+}
 
-  linux_entry "${OS}" "${version}" advanced \
-              "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
-  if [ "x${GRUB_DISABLE_RECOVERY}" != "xtrue" ]; then
-    linux_entry "${OS}" "${version}" recovery \
-                "${GRUB_CMDLINE_LINUX_RECOVERY} ${GRUB_CMDLINE_LINUX}"
-  fi
-done
+# Scan for a corresponding initrd image for the kernel.
+# $1: Full version of the kernel.
+get_initrd() {
+	local rd ret="${INITRD_EARLY}"
+	for rd in \
+		initramfs-"$1".img \
+		initramfs.img-"$1" \
+		initrd-"$1".img \
+		initrd.img-"$1" \
+		initrd-"$1".gz ; do
+		if [ -e "$KERNEL_DIR/$rd" ] ; then
+			info "$(gettext_printf "Initial ramdisk image for kernel %s: %s" "$1" "${GRUB_KERNEL_DIR}${rd}")"
+			ret="${ret} ${GRUB_KERNEL_DIR}${rd}"
+			break
+		else
+			continue
+		fi
+	done
+	printf "%s" "$ret"
+}
 
-# If at least one kernel was found, then we need to
-# add a closing '}' for the submenu command.
-if [ x"$is_top_level" != xtrue ]; then
-  echo '}'
-fi
+process_kernels() {
+	local kernel uname_r latest latest_kernels localpart
+	# Use the latest non-RC kernel.
+	# Whatever. I don't care.
+	IFS=$'\n'
+	latest="${KERNEL_VERSIONS[0]}"
+	latest_kernels=($(echo "${KERNELS[*]}" | grep "$latest"))
+	latest_kernels=($(prioritize ${latest_kernels[@]}))
+	unset IFS
+	ROOT_PARAM="root=$(find_rootdev)"
+	if [ -z "${ROOT_PARAM##root=}" ] ; then
+		gettext_printf "Internal error - Failed to find the root device\n" >&2
+		exit 1
+	fi
+	ARGS="${ROOT_PARAM} ${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
+	for kernel in "${latest_kernels[@]}" ; do
+		uname_r=${kernel/#@(vmlinu?|kernel)-/}
+		info $(gettext_printf "Processing kernel '%s'\n" "$uname_r")
+		localpart="${uname_r//$latest/}"
+		comments="$(get_comments "$uname_r")"
+		title="${OS}${comments}"
+		linux_entry "$title" "${GRUB_KERNEL_DIR}${kernel}" "$(get_initrd "$uname_r")" "$ARGS" "${OS_ID}-${kernel}"
+	done
+	# Now we processed the latest versions of the kernel, time to put
+	# other kernel versions into the submenu.
+	unset UNPROCESSED_VERSIONS["$latest"]
+	if [ "${#UNPROCESSED_VERSIONS[@]}" = "0" ] ; then
+		return
+	fi
+	submenu_title="$(gettext_printf "More kernel options for %s" "$OS")"
+	printf "submenu '%s' {\n" "$submenu_title"
+	indent="$indent$tab"
+	for version in "${UNPROCESSED_VERSIONS[@]}" ; do
+		IFS=$'\n'
+		kernels=($(echo "${KERNELS[*]}" | grep "$version"))
+		unset IFS
+		for kernel in "${kernels[@]}" ; do
+			uname_r=${kernel/#@(vmlinu?|kernel)-/}
+			info $(gettext_printf "Processing kernel '%s'\n" "$uname_r")
+			localpart="${uname_r//$latest/}"
+			comments="$(gettext_printf "With kernel %s" "$uname_r")"
+			comments=" ($(echo "$comments" | grub_quote))"
+			title="${OS}${comments}"
+			linux_entry "$title" "${GRUB_KERNEL_DIR}/${kernel}" "$(get_initrd "$uname_r")" "$ARGS" "${OS_ID}-${kernel}"
+		done
+	done
+	printf "}\n"
+}
 
-echo "$title_correction_code"
+scan_for_kernels
+process_kernels
+info "$(gettext_printf "Done generating entries for %s." "$OS")"
-- 
2.50.1

